miniapps/electromagnetics/maxwell_solver.cpp

```diff
@@ -0,0 +1,673 @@
+               const_cast<Array<int>&>(abc_marker_));
+         }
+      }
+
+      a1_[idt]->Assemble();
+      a1_[idt]->Finalize();
+      A1_[idt] = a1_[idt]->ParallelAssemble();
+
+      diagScale_[idt] = new HypreDiagScale(*A1_[idt]);
+      pcg_[idt] = new HyprePCG(*A1_[idt]);
+      pcg_[idt]->SetTol(1.0e-12);
+      pcg_[idt]->SetMaxIter(200);
+      pcg_[idt]->SetPrintLevel(0);
+      pcg_[idt]->SetPreconditioner(*diagScale_[idt]);
+   }
+}
+
+void
+MaxwellSolver::implicitSolve(double dt, const Vector &B, Vector &dEdt) const
+{
+   int idt = hCurlLosses_ ? ((int)(dtScale_ * dt / dtMax_)) : 0;
+
+   b_->Distribute(B);
+   weakCurlMuInv_->Mult(*b_, *rhs_);
+
+   if ( hCurlLosses_ )
+   {
+      e_->Distribute(*E_);
+      hCurlLosses_->AddMult(*e_, *rhs_, -1.0);
+   }
+
+   if ( jd_ )
+   {
+      jCoef_->SetTime(t); // 't' is member data from mfem::TimeDependentOperator
+      jd_->Assemble();
+      *rhs_ -= *jd_;
+   }
+
+   if ( dEdtBCCoef_ )
+   {
+      dedt_->ProjectBdrCoefficientTangent(*dEdtBCCoef_,
+                                          const_cast<Array<int>&>(dbc_marker_));
+   }
+
+   // Create objects and matrices for solving with the given time step
+   setupSolver(idt, dt);
+
+   // Apply essential BCs and determine true DoFs for the right hand side
+   a1_[idt]->FormLinearSystem(dbc_dofs_, *dedt_, *rhs_, *A1_[idt], dEdt, *RHS_);
+
+   // Solve for the time derivative of the electric field (true DoFs)
+   pcg_[idt]->Mult(*RHS_, dEdt);
+
+   // Distribute shared DoFs to relevant processors
+   a1_[idt]->RecoverFEMSolution(dEdt, *rhs_, *dedt_);
+}
+
+void
+MaxwellSolver::SyncGridFuncs()
+{
+   e_->Distribute(*E_);
+   b_->Distribute(*B_);
+}
+
+double
+MaxwellSolver::GetMaximumTimeStep() const
+{
+   if ( dtMax_ > 0.0 )
+   {
+      return dtMax_;
+   }
+
+   HypreParVector * v0 = new HypreParVector(HCurlFESpace_);
+   HypreParVector * v1 = new HypreParVector(HCurlFESpace_);
+   HypreParVector * u0 = new HypreParVector(HDivFESpace_);
+   HypreParVector * vTmp = NULL;
+
+   v0->Randomize(1234);
+
+   int iter = 0, nstep = 20;
+   double dt0 = 1.0, dt1 = 1.0, change = 1.0, ptol = 0.001;
+
+   // Create Solver assuming no loss operators
+   setupSolver(0, 0.0);
+
+   // Use power method to approximate the largest eigenvalue of the
+   // update operator.
+   while ( iter < nstep && change > ptol )
+   {
+      double normV0 = InnerProduct(*v0,*v0);
+      *v0 /= sqrt(normV0);
+
+      NegCurl_->Mult(*v0,*u0);
+      M2MuInv_->Mult(*u0,*HD_);
+      NegCurl_->MultTranspose(*HD_,*RHS_);
+
+      pcg_[0]->Mult(*RHS_,*v1);
+
+      double lambda = InnerProduct(*v0,*v1);
+      dt1 = 2.0/sqrt(lambda);
+      change = fabs((dt1-dt0)/dt0);
+      dt0 = dt1;
+
+      if ( myid_ == 0 && logging_ > 0 )
+      {
+         cout << iter << ":  " << dt0 << " " << change << endl;
+      }
+
+      vTmp = v0;
+      v0   = v1;
+      v1   = vTmp;
+
+      iter++;
+   }
+
+   delete v0;
+   delete v1;
+   delete u0;
+
+   return dt0;
```

Comment:
Given that the approximate nature of the power method, do we want a safety factor on the dt estimate?

Reply:
It wouldn't hurt.  In the main program I round dt down so that the final time is reached after an integer number of time steps.  This is not guaranteed to reduce dt at all let alone by a sufficient amount.  What's your favorite factor?  I usually pick something in the ball park of 90-95%.

Reply:
Actually, we can incorporate this into the rounding routine which is called from main.   Consequently, we can easily make this factor user defined.

Reply:
Yeah, that's probably best.  Unless you can get a good error bound on the eigenvalue estimate, but my guess is that will require an estimate of the second largest eigenvalue which typically drives the power method convergence.